### PR TITLE
DietPi-DDNS | Manage domains for your dynamic IP

### LINKED ISSUE #2868: DietPi-Software | DuckDNS
#### Is your feature request related to a problem? Please describe:

- duckdns is an easy and free to use alternative to no-ip,that can be asily setup from terminal

#### Describe the solution you'd like:

- could be added to the software tab as an alternative to the already existent softwares

#### Additional context

- it dows exactaly the same as no-ip,so why not put it in there since it is free,and simple to setup? 
the own duckdns website already gives the bash script to setup in solutions like the dietpi-software (just change the variables by the given website name and the given token)

#### Vote for this feature on https://feathub.com/MichaIng/DietPi/+85

### PR DESCRIPTION
**Pull request by nonnorm**
**Status**: Review

**References**:
- https://github.com/MichaIng/DietPi/issues/2868
- https://dietpi.com/phpbb/viewtopic.php?t=8041

**ToDo**:
- [x] Replace inline cURL with dedicated script to avoid having plain text passwords/tokens in journalctl logs
- [x] Replace password input box with password box (which replaces typed characters with `*`)? Enhances security but especially with for those long tokens it's a comfort downside. ... Ah, it's the very same with UNIX passwords, let's be consequent here. Via SSH copy&paste is always possible anyway.

### CODE REVIEW

**Comment by MichaIng** [comment]
Ay, before you invest too much time into the software install part of this, I'm working on a DietPi-DDNS script with generic DDNS provider support for those which can be updated via simple request, including support to enter a custom provider update URL. Similarly to DietPi-VPN, there is not much point in having this as a software install option as the script will be always present.

I'll incorporate/merge it with our DuckDNS specific work, especially the "OK" return code to check for success is a good idea.

**Comment by MichaIng** [comment]
I restored the branch, and re-opened PR to compare/incorporate the work.

**Comment by nonnorm** [comment]
It's ready for you, @MichaIng.

File: dietpi/dietpi-config [hunk 1]
```diff
@@ -4190,6 +4190,20 @@ Additional benchmarks:
...
+		# DuckDNS
+		local duckdns_installed=0
+		local duckdns_menutext='[Not Installed] Install DuckDNS now'
+		if [[ -f '/var/lib/dietpi/dietpi-software/installed/duckdns.sh' ]]; then
+
+			duckdns_installed=1
+			duckdns_menutext='[Offline] Invalid DuckDNS details'
+
+			# Check if DuckDNS url returns OK
+			[[ "$(eval $(grep 'duckdns.org/update' /var/lib/dietpi/dietpi-software/installed/duckdns.sh))" == 'OK' ]] && duckdns_menutext='[Online] Change DuckDNS details'
```
Comment (MichaIng) [inline dietpi/dietpi-config]: @ravenclaw900 
Little not about the CodeFactor error here:
- The left sided double quotes in bash `[[ ]]` is not required, as word splitting doesn't play a role and glob expansion (like allowing `*` and `[0-9]`) is only done on unquoted patterns on the right side of `==`.
- But inside of command substitutions `$(command arguments...)`, word splitting is relevant as with every other command call.
- This is particular case, `eval` will to execute the intended `curl`, as `grep` returns multiple arguments, while eval needs to get everything in a single argument. So the statement needs to be:
    ```
    [[ $(eval "$(grep 'duckdns.org/update' /var/lib/dietpi/dietpi-software/installed/duckdns.sh)") == 'OK' ]]
    ```

But do not bother, I'm adding `dietpi-ddns` after doing another important change to `dietpi-software` from my ToDo.

**Comment by MichaIng** [comment]
To cover: [NextDNS](https://feathub.com/MichaIng/DietPi/+167)

**Comment by Joulinar** [comment]
Question: what happen with user who have NoIP installed as of today. Are they able to continue to mange their NoIP account? Or do they need to re-setup thinks using `dietpi-ddns` or are we going to provide a migration path? I mean we are going to remove the install option, means people are not able to uninstall via `dietpi-software` anymore. Probably I missed something as I was checking the code just briefly

**Comment by MichaIng** [comment]
AFAIK `/usr/local/etc/no-ip.conf` contains the password in a hashed form, so there is no way for us to do an automated migration. The No-IP client stays installed and functional. Only when using `dietpi-ddns` to setup and finally apply a new DDNS Cron job, and the update test succeeds, the No-IP client + service is removed. The setup would need to be done manually indeed.

What we should add is an interactive prompt to inform and encourage users with No-IP installed to setup No-IP DDNS via our new script, which is lighter and uses up-to-date protocols etc and have the old client automatically removed as very last step of the setup. Most importantly it cannot happen that the client is removed but no Cron job installed.

**Comment by MichaIng** [comment]
That should make it clear: https://github.com/MichaIng/DietPi/pull/4229/commits/5c83d4eed7bb5f56c510ad1511345c762acc842e

**Comment by Joulinar** [comment]
Yeah that should be fine 😉

**Review by MichaIng — DISMISSED** [review]

**Review by MichaIng — DISMISSED** [review]

File: dietpi/dietpi-ddns [hunk 2]
```diff
@@ -80,7 +80,7 @@ Read()
 		[[ $PASSWORD ]] || PASSWORD=${cronjob#*token=} PASSWORD=${PASSWORD%\'*}
 
 	# No-IP
-	elif [[ $cronjob == *'no-ip.com'* ]]
+	elif [[ $cronjob == *'noip.com'* ]]
```
Comment (MichaIng) [inline dietpi/dietpi-ddns]: @ravenclaw900 
The API docs say "no-ip.com": https://www.noip.com/integrate/request
But indeed both work. They are pointing to the same IP, just slightly different TTL:
- https://dnschecker.org/all-dns-records-of-domain.php?query=dynupdate.no-ip.com&rtype=ALL&dns=cloudflare
- https://dnschecker.org/all-dns-records-of-domain.php?query=dynupdate.noip.com&rtype=ALL&dns=cloudflare

File: dietpi/dietpi-ddns [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+# Apply chosen settings and create a Cron job for dynamic DNS updates
+Apply()
+{
+	# Generate URL
+	local url=$PROVIDER
+	# - DuckDNS
+	if [[ $PROVIDER == 'DuckDNS' ]]
+	then
+		url="https://www.duckdns.org/update?domains=$DOMAINS&token=$PASSWORD"
+		unset -v USERNAME PASSWORD
```
Comment (nonnorm) [inline dietpi/dietpi-ddns:120]: > unset -v USERNAME PASSWORD

Any reason for this line?  Just asking because it makes the 'Token' field disappear.

Comment (MichaIng) [reply dietpi/dietpi-ddns:120]: Ah, I didn't think about the values shown in the menu afterwards. Currently this is required to not have HTTP authentication used.

But then we need to a separate variable for this:
```
local http_auth=1
...
if [[ $PROVIDER == 'DuckDNS' ]]
then
	unset -v http_auth
...
...${http_auth:+ -u '$USERNAME:$PASSWORD'}...
```
Will implement this now.

Comment (MichaIng) [reply dietpi/dietpi-ddns:120]: Done: https://github.com/MichaIng/DietPi/commit/5ce84c7ed90ffd6c882e18ee2f2f3c2854647838

**Pull request merged**